lua/luasnip/extras/fmt.lua

```diff
@@ -0,0 +1,203 @@
+local function escape_set_char(c)
+    return (c == ']' or c == '%') and ('%'..c) or c
+end
+
+-- Interpolate elements from `args` into format string with placeholders.
+--
+-- The placeholder syntax for selecting from `args` is similar to fmtlib and
+-- Python's .format(), with some notable differences:
+-- * no format options (like `{:.2f}`)
+-- * 1-based indexing
+-- * numbered/auto-numbered placeholders can be mixed; numbered ones set the
+--   current index to new value, so following auto-numbered placeholders start
+--   counting from the new value (e.g. `{} {3} {}` is `{1} {3} {4}`)
+--
+-- Arguments:
+--   fmt: string with placeholders
+--   args: table with list-like and/or map-like keys
+--   opts:
+--     delimiters: string, 2 distinct characters (left, right), default "{}"
+--     strict: boolean, set to false to allow for unused `args`, default true
+-- Returns: a list of strings and elements of `args` inserted into placeholders
+local function interpolate(fmt, args, opts)
+    local defaults = {
+        delimiters = '{}',
+        strict = true,
+    }
+    opts = vim.tbl_extend('force', defaults, opts or {})
+
+    -- sanitize delimiters
+    assert(#opts.delimiters == 2,
+        'Currently only single-char delimiters are supported, e.g. delimiters="{}" (left, right)')
+    assert(opts.delimiters:sub(1, 1) ~= opts.delimiters:sub(2, 2),
+        "Delimiters must be two _different_ characters")
+    local delimiters = {
+        left = opts.delimiters:sub(1, 1),
+        right = opts.delimiters:sub(2, 2),
+        esc_left = escape_set_char(opts.delimiters:sub(1, 1)),
+        esc_right = escape_set_char(opts.delimiters:sub(2, 2)),
+    }
+
+    -- manage insertion of text/args
+    local elements = {}
+    local last_index = 0
+    local used_keys = {}
+
+    local add_text = function(text)
+        if #text > 0 then
+            table.insert(elements, text)
+        end
+    end
+    local add_arg = function(placeholder)
+        local num = tonumber(placeholder)
+        local key
+        if num then  -- numbered placeholder
+            last_index = num
+            key = last_index
+        elseif placeholder == '' then  -- auto-numbered placeholder
+            key = last_index + 1
+            last_index = key
+        else  -- named placeholder
+            key = placeholder
+        end
+        assert(args[key], string.format('Missing key `%s` in format arguments: `%s`', key, fmt))
+        table.insert(elements, args[key])
+        used_keys[key] = true
+    end
+
+    -- iterate keeping a range from previous match, e.g. (not in_placeholder vs in_placeholder)
+    -- "Sample {2} string {3}."   OR  "Sample {2} string {3}."
+    --       left^--------^right  OR      left^-^right
+    local pattern = string.format('[%s%s]', delimiters.esc_left, delimiters.esc_right)
+    local in_placeholder = false
+    local left = 0
+
+    while true do
+        local right = fmt:find(pattern, left + 1)
```

Comment:
What do you think of doing a plain-text search for 
- `delim_left`, if in_placeholder and
- `delim_right`, if not in_placeholder ?
That shouldn't be slower and be a bit simpler (no need to construct a set from escaped delimiters)

Reply:
That would also require adjusting the un-escape code, ie. only skipping escaped matches and actually replacing eg. `{{` with `{` in `add_text`

Reply:
I would rather stick with this simple regex search for now. I feel like doing plain text search would unnecessarily complicate the escaping logic. I guess it may even make things slower, because on each `text_add` we would need to `gsub` all the `}}` if I'm correct.

Reply:
True, it's alright if we leave it like this👍